### ISSUE TITLE
Wrong order of entries can produce uninstallable RPMs

### ISSUE THREAD

**Issue by oreinert** [ctron/rpm-builder#14]
Consider the example in the following archive:
[pluginbug.zip](https://github.com/ctron/rpm-builder/files/1000859/pluginbug.zip)

It contains sources for building a simple RPM containing:
 * the directory `/opt/test`, and
 * the file `/opt/test/README.md`.

There's a `pom.xml` file for building the RPM via the Maven RPM plugin, and a .spec file for building the same RPM using the `rpmbuild` command (by running `rpmbuild --define="%_topdir $PWD" -bb SPECS/rpm-plugin-test.spec`).

In the example, `%dir /opt/test` is specified _after_ the file, and this causes the Maven plugin to generate an RPM that can't be installed:

```
$ sudo rpm -i target/rpm-plugin-test-0.1.0-1-noarch.rpm
error: unpacking of archive failed on file /opt/test/README.md;59198b61: cpio: open failed - No such file or directory
error: rpm-plugin-test-0.1.0-1.noarch: install failed
```

The same RPM generated using `rpmbuild` installs without issue. Looking at the entries in the cpio archive within the two RPMs seems revealing:

```
$ rpm2cpio target/rpm-plugin-test-0.1.0-1-noarch.rpm | cpio -tv
-rw-r--r--   1 root     root           48 May 15 12:51 ./opt/test/README.md
drwxr-xr-x   2 root     root            0 May 15 12:51 ./opt/test
1 block
$ rpm2cpio RPMS/noarch/rpm-plugin-test-0.1.0-0.noarch.rpm | cpio -tv
drwxr-xr-x   2 root     root            0 May 15 12:51 ./opt/test
-rw-r--r--   1 root     root           48 May 15 12:51 ./opt/test/README.md
1 block
```

The problem seems to be that the RPM generated by the plugin fails to install because cpio can't write the README.md file because the folder it should be in doesn't exist yet. The Maven plugin should order entries in the archive in such a way that it can always be installed, like the `rpmbuild` command.

Fixing the given example is trivial, of course - just swap the order of the entries. However, for more complicated builds, this may not be possible, especially when profiles are used to contribute independent parts of the RPM plugin config using stuff like http://blog.sonatype.com/2011/01/maven-how-to-merging-plugin-configuration-in-complex-projects/ .

**Comment by ctron** [comment]
Yes it is true, this condition is not being checked. And I agree, although there is a workaround for that, there should be some proper solution.

**Comment by dwalluck** [comment]
Fixed via #90.

**Comment by dwalluck** [comment]
I couldn't reproduce using the example. The order is correct, even though the order in the pom itself is reversed:

```bash
$ rpm -qpl rpm-plugin-test-0.1.0-1-noarch.rpm
/opt/test
/opt/test/README.md
```

**Comment by moedan** [comment]
I still can reproduce the issue mentioned by @oreinert with version 1.13.0:

```
rpm2cpio target/rpm-plugin-test-0.1.0-1.noarch.rpm | cpio -tv
-rw-r--r--   1 root     root           49 Apr 22 08:37 ./opt/test/README.md
drwxr-xr-x   2 root     root            0 Apr 22 08:37 ./opt/test
```


I think the output of cmd `rpm -qpl target/rpm-plugin-test-0.1.0-1.noarch.rpm` is misleading here. Maybe the output is always sorted, but I have no evidence for this thesis.

#90 fixes a problem with entry ordering only in context of the "generate intermediate directories" feature.

As mentioned in #90, there is an idea to solve the problem with the entry order in a universal way by tacking the order of the added entries to the RPM during the build process. In case of an invalid sequence, the build process should be cancelled. Unfortunately, I haven't had the time yet to implement it.

**Comment by dwalluck** [comment]
I didn't look further, but it installs on the most recent Fedora, so something has created the directory. I think this is showing that RPM itself must be handling the ordering at install time rather than build time. I think RPM Builder would need to handle it during unpacking. But, let me know if you want me to reopen the issue.

**Comment by moedan** [comment]
I've verified the installation on Red Hat Universal Base Image 8.9 and 9.5. It seems to work as you described. So it is not neccessary for me to reopen the issue.

**Comment by dwalluck** [comment]
There may still me an issue with the rpm unpack mojo. So it may work fine in the real rpm, but I will have to make sure that the Java version can also unpack it.
